### ISSUE TITLE
Incomplete download

### ISSUE THREAD

**Issue by Piero01** [warmcat/libwebsockets#1229]
Hello, since commit c60b241 of Aug 21, 2017 I get problem of incomplete downloads with relatively big files (> 100 Mo). After some time the download fails and I need to reload several times before completing the entire file.

**Comment by lws-team** [comment]
The commit you reference is about time that an ah is held.

But the issue is about download of a "large file".  So I guess when it fails, what is actually happening is the timeout due to an ah being held for what lws thinks is an unreasonable about of time.

Can you let me know the version of lws you tried... did you try current master?

Can you turn debug logging on and capture the log when the download fails, to make sure it's what we think it is?

On master, "excessive ah hold" is considered `ah->wsi->vhost->timeout_secs_ah_idle + 360`, ie, whatever you set in `info.timeout_secs_ah_idle` (in secs) plus 5 minutes.  So you can control this behaviour actually.

**Comment by lws-team** [comment]
Thanks.  But the point is you can control the `timeout_secs_ah_idle` part of "timeout_secs_ah_idle + 5min", at context creation time.

https://github.com/warmcat/libwebsockets/blob/master/lib/libwebsockets.h#L2973

> Maybe there is another problem. With the stable branch I can continue the download 

On master you need to configure with `-DLWS_WITH_RANGES=1` if you want rfc7233 http ranges, the default has changed to disabled.

**Comment by Piero01** [comment]
Thank you very much for your reactivity. I put a timeout of 1 hour to be sure but I don't like this solution very much. I'm not sure to know what ah is but it seems strange to have a timeout for unused connection while the connection should still be active to finish the download. Nevertheless it's not a big issue.

> On master you need to configure with -DLWS_WITH_RANGES=1 if you want rfc7233 http ranges, the default has changed to disabled.

Thanks for the information it will avoid me to loose a lot of time later. Is there some place where I can find what to do to keep backward compatibility when upgrading version of libwebsocket?

**Comment by lws-team** [comment]
> it seems strange to have a timeout for unused connection

If you consider lws being used a public-facing service, you do not want users to tie up all your resources cheaply and DoS you.

> Is there some place where I can find what to do to keep backward compatibility when upgrading version of libwebsocket?

I update the changelog before every release

https://github.com/warmcat/libwebsockets/blob/master/changelog

but master is not yet released, so eyeballing the git log is a good way; that's how I produce the changelog

```
$ git log v2.4.0..
...
commit 5b74d7108b5cfcc3a6ee3dcad521e0f0c3a18d0c
Author: Andy Green <andy@warmcat.com>
Date:   Tue Nov 14 11:25:54 2017 +0800

    gcov: add in cmake and adapt defaults
    
    1) Introduce LWS_WITH_GCOV to build with gcc / clang coverage instrumentation.
    
    $ cd build
    $ make clean && rm -f `find . -name "*.gcno" -o -name "*.gcda"` && make -j16 && sudo make install && sudo /usr/local/bin/libwebsockets-test-server -s
    ...
    $ gcov `find . -name *.c.gcno | grep -v test-apps` -b | sed "/\.h.\$/,/^$/d"
    
    The above are available in two helper scripts
    
     - scripts/build-gcov.sh
     - scripts/gcov.sh
    
    2)
    
    CMake defaults changed:
    
     - LWS_WITH_ZIP_FOPS: OFF
     - LWS_WITH_RANGES: OFF
     - LWS_WITHOUT_EXTENSIONS: ON
     - LWS_WITH_ZLIB: OFF
    
    New CMake controls that default-OFF:
    
     - LWS_WITH_GENHASH
     - LWS_WITH_GENRSA
    
    these are implied by LWS_WITH_JWS (which is implied by LWS_WITH_ACME)
    
    3) rename ./lib/tls/XXX/server.c and client.c to XXX-server.c / XXX-client.c.
    
    This is because gcov dumps its results using the .c filename part only,
    the copies overwrite each other if there are different .c files in the tree
    with the same filename part.
    
    4) Add onetime test-client mode and test to ./test-apps/attack.sh
    
    5) Add gcov howto in READMEs/README.build.md using attack.sh
...
```

**Comment by Piero01** [comment]
Yes, you put the finger on my problem. I'm concerned about DoS. You cut my sentence in the middle, maybe I was not clear. I'm not asking to remove the timeout. I totally agree that we need a **short** timeout on unused connection. But the problem in this case is that it is not an unused connection and the timeout should be set enough to keep alive a connection that is still in use.
I'm forced to put a big timeout (and be more prone to DoS) in order to allow download of big files without trouble. 
I think the timeout should not occur when the connection is still necessary (for downloading a file for exemple). I don't know if it is possible to take into account of the timer only when the connection is really no more in use.

My dream is to keep a short timeout and to be able to download files without a timeout occur.

NB: As a side effect, because of the high timeout value, I also was forced to rise the _max_http_header_pool_ to accept more clients in parallel.

**Comment by lws-team** [comment]
The situation is as it is for reasons.

The original ur-lws had the ah as a permanent part of the wsi.  But for ws-upgraded wsi, who are long-lived and no longer use http and have no use for headers, that is a huge waste of space since each ah is a few KiB.  If your use case is mainly ws serving, that's quite a bloat.

So ah were split out to be dynamically allocated from a pool that was allocated at context creation time.  That was pretty reasonable, if you have long-lived http connections new connections would wait for a free ah.  It worked efficiently with http/1.1 pipelining, the wsi kept the single ah to do the pipelined requests serially.  I had plans to drop the ah on long-lived http connections after the HTTP callback, and before POST payloads, but before then I added http/2 support.

H2 has a completely different pattern of use for the ah.  Modern browsers may spawn eg 16 or more streams inside the h2 connection rapidly one after the other; it requires 16 ah.  H2 header processing cannot be deferred and done out of order, not only because blocking read blocks all streams but because earlier headers change the hpack state used to encode later ones.  So usefully deploying h2 puts us back in the situation you may need a lot of ah dynamically.

So that brings us to today, there is an ah pool but they are allocated on-demand and freed when idle.

The addition checking for ah held an 'unreasonable time' is to track down cases of ah pool exhaustion I saw on my public server... that turned out to be a broken h1 bot that never completed the http action nor closed the connection nor otherwise timed out.

Different use-cases for lws can't avoid some tuning of the ah stuff...
